apollo-router/src/services/layers/persisted_queries/mod.rs

```diff
@@ -184,57 +150,84 @@ impl PersistedQueryLayer {
         }
     }
 
-    /// Handles incoming freeform GraphQL requests according to the safelisting configuration options
-    pub(crate) fn handle_freeform_graphql(
+    pub(crate) async fn supergraph_request_with_analyzed_query(
         &self,
         request: SupergraphRequest,
-        manifest_poller: Arc<PersistedQueryManifestPoller>,
     ) -> Result<SupergraphRequest, SupergraphResponse> {
-        if let Some(operation_body) = request.supergraph_request.body().query.as_ref() {
-            // TODO: replace this with proper introspection parsing
-            if self.sandbox_enabled
-                && (operation_body == SANDBOX_INTROSPECTION_QUERY_WO_DEPRECATED_ARGS
-                    || operation_body == SANDBOX_INTROSPECTION_QUERY_W_DEPRECATED_ARGS)
-            {
-                // if sandbox is enabled and the incoming operation is the introspection query sent by sandbox,
-                // allow the request to continue on.
+        let manifest_poller = match &self.manifest_poller {
+            // PQ feature entirely disabled; just pass through.
+            None => return Ok(request),
+            Some(mp) => mp,
+        };
+
+        let operation_body = match request.supergraph_request.body().query.clone() {
+            // if the request doesn't have a `query` document, continue with normal execution, which
+            // will result in the normal no-operation error.
+            None => return Ok(request),
+            Some(ob) => ob,
+        };
+
+        let compiler = {
+            let context_guard = request.context.private_entries.lock();
+
+            if context_guard.get::<UsedQueryIdFromManifest>().is_some() {
+                // We got this operation from the manifest, so there's no
+                // need to check the safelist.
+                drop(context_guard);
                 return Ok(request);
             }
 
-            let mut is_persisted = None;
-
-            if self.log_unknown
-                && !is_operation_persisted(
-                    &mut is_persisted,
-                    manifest_poller.clone(),
-                    operation_body,
-                )
-            {
-                tracing::warn!(message = "unknown operation", operation_body);
+            match context_guard.get::<Compiler>() {
+                None => {
+                    drop(context_guard);
+                    // For some reason, QueryAnalysisLayer didn't give us a Compiler?
+                    return Err(supergraph_err(
+                        graphql_err(
+                            "MISSING_PARSED_OPERATION",
+                            "internal error: compiler missing",
+                        ),
+                        request,
+                        ErrorCacheStrategy::DontCache,
+                    ));
```

Comment:
We tend to rely on https://doc.rust-lang.org/std/ops/enum.ControlFlow.html to differenciate between errors that will yield an http response, (ControlFlow::Break) and BoxErrors (HTTP 500).

This looks like this should be a BoxError. Did we consider using Result<ControlFlow<SupergraphRequest, SupergraphResponse>, BoxError>> ?

Reply:
I'm not too familiar with that enum, and it's not currently used anywhere in this layer or its friends like the APQ layer or query analysis layer. Is the idea that I would make the top level entry points to the layer (`supergraph_request` etc) return this enum and then update the code in `router_service.rs` to process its output?

Reply:
I see, that's ok then. I tend to rely on controlflow when i need to differenciate between errors I want to reply to the client, and router errors that will bubble into a 500.

`compiler missing` looks like something a client should never see.

Reply:
ah that checks out. probably worth doing then! what do you think about failure mode for being unable to acquire locks? currently it's a panic.

Reply:
FWIW I modeled this after BridgeQueryCompiler::call which does what looks like the same thing as a parse error if this object is missing... Not sure what change to make here.

Reply:
> what do you think about failure mode for being unable to acquire locks? currently it's a panic.
I think this is one of the rare cases where a panic would be good. I'm not sure how to recover from a poisoned lock.

> FWIW I modeled this after BridgeQueryCompiler::call which does what looks like the same thing as a parse error if this object is missing... Not sure what change to make here.

I see, then there's probably a bit of work for us to do in both layers, would you mind opening a followup issue?